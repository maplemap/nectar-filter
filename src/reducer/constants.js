export const REMOVE_APPLIED_FILTER_ITEM = 'REMOVE_APPLIED_FILTER_ITEM';
export const SET_APPLIED_FILTERS = 'SET_APPLIED_FILTERS';
export const REMOVE_APPLIED_FILTER_TYPE = 'REMOVE_APPLIED_FILTER_TYPE';
export const CLEAR_APPLIED_FILTERS = 'CLEAR_APPLIED_FILTERS';
