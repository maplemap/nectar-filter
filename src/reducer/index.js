export {ContextApp, initialState, reducer} from './reducer';
