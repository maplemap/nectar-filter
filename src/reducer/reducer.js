import React from 'react';
import {isMobile} from 'react-device-detect';
import filtersData from 'data';
import {
    REMOVE_APPLIED_FILTER_ITEM,
    SET_APPLIED_FILTERS,
    CLEAR_APPLIED_FILTERS, REMOVE_APPLIED_FILTER_TYPE
} from './constants';

export const ContextApp = React.createContext({});

export const initialState = {
    filtersData,
    isMobile,
    appliedFilters: {}
};

export const reducer = (state = {}, action = {}) => {
    const {type, payload} = action;

    // eslint-disable-next-line default-case
    switch (type) {
        case SET_APPLIED_FILTERS: {
            const {checkedFiltersIds, filterType} = payload;

            return {
                ...state,
                appliedFilters: {
                    ...state.appliedFilters,
                    [filterType]: checkedFiltersIds
                }
            }
        }

        case REMOVE_APPLIED_FILTER_TYPE: {
            const {filterType} = payload;
            const newAppliedFilters = {...state.appliedFilters};
            delete newAppliedFilters[filterType];

            return {
                ...state,
                appliedFilters: newAppliedFilters
            }
        }

        case REMOVE_APPLIED_FILTER_ITEM: {
            const {filterType, filter} = payload;
            const {appliedFilters} = state;
            const newFiltersByFilterType = appliedFilters[filterType].filter(item => item !== filter);
            const newAppliedFilters = {...appliedFilters};
            if (newFiltersByFilterType.length === 0) {
                delete newAppliedFilters[filterType];
            } else {
                newAppliedFilters[filterType] = newFiltersByFilterType;
            }

            return {
                ...state,
                appliedFilters: newAppliedFilters
            }
        }

        case CLEAR_APPLIED_FILTERS: {
            return {
                ...state,
                appliedFilters: {}
            }
        }
    }

    return state;
};
