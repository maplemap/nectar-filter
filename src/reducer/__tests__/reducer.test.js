import {
    SET_APPLIED_FILTERS,
    REMOVE_APPLIED_FILTER_TYPE,
    REMOVE_APPLIED_FILTER_ITEM,
    CLEAR_APPLIED_FILTERS
} from '../constants';
import {reducer} from '../reducer';

describe('reducer', () => {
    it('should return `state`', () => {
        const state = {};

        expect(reducer(state)).toEqual(state);
    });

    describe('when action `type` is equal to `SET_APPLIED_FILTERS`', () => {
        it('should set `appliedFilters` by `filterType`', () => {
            const checkedFiltersIds = ['filter'];
            const filterType = 'filterType';
            const action = {
                type: SET_APPLIED_FILTERS,
                payload: {checkedFiltersIds, filterType}
            };

            expect(reducer({}, action)).toEqual({
                appliedFilters: {[filterType]: checkedFiltersIds}
            });
        });
    });

    describe('when action `type` is equal to `REMOVE_APPLIED_FILTER_TYPE`', () => {
        it('should remove key by `filterType` in `appliedFilters`', () => {
            const filterType = 'filterType';
            const state = {
                appliedFilters: {
                    [filterType]: ['filter'],
                    filterType2: ['filter2']
                }
            };
            const action = {
                type: REMOVE_APPLIED_FILTER_TYPE,
                payload: {filterType}
            };
            reducer(state, action);

            expect(reducer(state, action)).toMatchSnapshot();
        });
    });

    describe('when action `type` is equal to `REMOVE_APPLIED_FILTER_ITEM`', () => {
        it('should return `appliedFilters` without removed items in `filterType`', () => {
            const filterType = 'filterType';
            const filter = 'filter1';
            const state = {
                appliedFilters: {
                    [filterType]: [filter, 'filter2']
                }
            };
            const action = {
                type: REMOVE_APPLIED_FILTER_ITEM,
                payload: {filterType, filter}
            };

            expect(reducer(state, action)).toMatchSnapshot();
        });
    });

    describe('when action `type` is equal to `CLEAR_APPLIED_FILTERS`', () => {
        it('should return `appliedFilters` like a empty object', () => {
            const state = {
                appliedFilters: {
                    filterType: ['filter1', 'filter2']
                }
            };
            const action = {
                type: CLEAR_APPLIED_FILTERS
            };

            expect(reducer(state, action)).toEqual({appliedFilters: {}});
        });
    });
});
