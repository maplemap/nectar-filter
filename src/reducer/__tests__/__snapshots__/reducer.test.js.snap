// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer when action \`type\` is equal to \`REMOVE_APPLIED_FILTER_ITEM\` should return \`appliedFilters\` without removed items in \`filterType\` 1`] = `
Object {
  "appliedFilters": Object {
    "filterType": Array [
      "filter2",
    ],
  },
}
`;

exports[`reducer when action \`type\` is equal to \`REMOVE_APPLIED_FILTER_TYPE\` should remove key by \`filterType\` in \`appliedFilters\` 1`] = `
Object {
  "appliedFilters": Object {
    "filterType2": Array [
      "filter2",
    ],
  },
}
`;
